### PR TITLE
[Google-Cloud/Storage] fix GCS types to work at runtime 

### LINKED ISSUE #17818: @types/google-cloud__storage doesn't actually work
- [x] I tried using the `@types/xxxx` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @blove @mlegenhausen

The export of the @google-cloud/storage module is the Storage object itself, not a namespace. 

So if you do:

```
import Storage = require('@google-cloud/storage');
```

and then try to create an instance like:

```
new Storage({ ... })
```

it fails because the types have it exporting a namespace of types, not the Storage object itself.

Also note that the test is all wrong here, it imports a bunch of stuff that doesn't actually exist. It passes given the types, but doesn't match what the library does at all:

https://github.com/blove/DefinitelyTyped/blob/c3d08f7d2e4473b36cae15ebfa82cd442ce1e0fa/types/google-cloud__storage/google-cloud__storage-tests.ts

```
import {
  // ...
  Storage
} from "@google-cloud/storage";
```

but Storage here is undefined with the real @google-cloud/storage code. Since it exports the constructor function for "Storage", with expandos for the other types.

### PR DESCRIPTION
**Pull request by baronfel**
We don't export a constructor, instead we export a namespace + function combo that works.

This fixes #17818.  I've tested this with some components at work and am able to successfully list files and whatnot.

I tried to keep the individual files for clarity, but I couldn't do that and make it work. For each file I tried the 'namespace with types + exported function' combo that we do for the `Storage` type here, but I couldn't re-export the namespaces in the root.

- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [X] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [X] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/DefinitelyTyped/DefinitelyTyped/issues/17818
- [X] Increase the version number in the header if appropriate.
- [X] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

File: types/google-cloud__storage/index.d.ts [hunk 1]
```diff
@@ -5,7 +5,339 @@
...
+    class File {
+        constructor(bucket: Bucket, name: string, options?: BucketFileOptions);
+        acl: Acl;
+        copy(destination: string | Bucket | File): Promise<[File, ApiResponse]>;
+        createReadStream(options?: ReadStreamOptions): ReadStream;
+        createWriteStream(options?: WriteStreamOptions): WriteStream;
+        delete(): Promise<[ApiResponse]>;
+        download(options?: DownloadOptions): Promise<[string]>;
+        exists(): Promise<[boolean]>;
+        get(): Promise<[File, ApiResponse]>;
```
Comment (ento) [inline types/google-cloud__storage/index.d.ts]: Not a blocker (this PR leaves things better than how it is now): [the resolved `File` object will have `metadata` property](https://github.com/GoogleCloudPlatform/google-cloud-node/blob/storage-1.2.0/packages/storage/src/file.js#L1146).

I think we want

```typescript
metadata?: FileMetadata
```

to capture that.

File: types/google-cloud__storage/index.d.ts [hunk 2]
```diff
@@ -5,7 +5,339 @@
...
+     * A file in the cloud.
+     */
+    class File {
+        constructor(bucket: Bucket, name: string, options?: BucketFileOptions);
+        acl: Acl;
+        copy(destination: string | Bucket | File): Promise<[File, ApiResponse]>;
+        createReadStream(options?: ReadStreamOptions): ReadStream;
+        createWriteStream(options?: WriteStreamOptions): WriteStream;
+        delete(): Promise<[ApiResponse]>;
+        download(options?: DownloadOptions): Promise<[string]>;
```
Comment (ento) [inline types/google-cloud__storage/index.d.ts]: Not a blocker (this PR leaves things better than how it is now): [File.download gives back an array of buffers.](https://github.com/GoogleCloudPlatform/google-cloud-node/blob/storage-1.2.1/packages/storage/src/file.js#L1008), not strings.

File: types/google-cloud__storage/index.d.ts (hunk 1 above)
Comment (baronfel) [inline types/google-cloud__storage/index.d.ts]: Thanks! Updated to reflect this.

File: types/google-cloud__storage/index.d.ts (hunk 2 above)
Comment (baronfel) [inline types/google-cloud__storage/index.d.ts]: Thanks! Updated to reflect this.

File: types/google-cloud__storage/index.d.ts [hunk 3]
```diff
@@ -5,7 +5,340 @@
...
+        get(options?: BucketGetOptions): Promise<[Bucket, Storage.ApiResponse]>;
+        getFiles(query?: BucketQuery): Promise<[File[]]>;
+        getFilesStream(query?: BucketQuery): ReadStream;
+        getMetadata(): Promise<[BucketMetadata, Storage.ApiResponse]>;
+        id: string;
+        makePrivate(options?: BucketPrivacyOptions): Promise<[File[]]>;
+        makePublic(options?: BucketPrivacyOptions): Promise<[File[]]>;
+        metadata: BucketMetadata;
+        name: string;
+        setMetadata(metadata?: BucketMetadata): Promise<[Storage.ApiResponse]>;
```
Comment (ento) [inline types/google-cloud__storage/index.d.ts]: Also takes an `OperationOptions` object.

Comment (ento) [inline types/google-cloud__storage/index.d.ts]: [`setLabels`](https://googlecloudplatform.github.io/google-cloud-node/#/docs/storage/1.2.0/storage/bucket?method=setLabels) seems to be missing.

File: types/google-cloud__storage/index.d.ts [hunk 4]
```diff
@@ -5,7 +5,340 @@
...
+        end?: number;
+        start?: number;
+        validation?: string | boolean;
+    }
+
+    /**
+     * Options when uploading file to bucket.
+     */
+    interface UploadOptions extends WriteStreamOptions {
+        destination?: string;
```
Comment (ento) [inline types/google-cloud__storage/index.d.ts]: [Takes quite a lot more options](https://googlecloudplatform.github.io/google-cloud-node/#/docs/storage/1.2.0/storage/bucket?method=upload).

File: types/google-cloud__storage/index.d.ts [hunk 5]
```diff
@@ -5,7 +5,340 @@
...
+        getBucketsStream(query?: BucketQuery): Promise<[ReadStream]>;
+        Channel: (storage: Storage, id: string, resourceId: string) => Channel;
+        File: (bucket: Bucket, name: string, opts: BucketFileOptions) => File;
+        Bucket: (storage: Storage, name: string) => Bucket;
+    }
+
+    /**
+     * This class allows you interact with Google Cloud Storage.
+     */
+    class Channel {
```
Comment (ento) [inline types/google-cloud__storage/index.d.ts]: <3 simple classes.

File: types/google-cloud__storage/index.d.ts [hunk 6]
```diff
@@ -5,7 +5,340 @@
...
+        min?: number;
+    }
+
+    /**
+     * A file in the cloud.
+     */
+    class File {
+        constructor(bucket: Bucket, name: string, options?: BucketFileOptions);
+        acl: Acl;
+        copy(destination: string | Bucket | File): Promise<[File, ApiResponse]>;
```
Comment (ento) [inline types/google-cloud__storage/index.d.ts]: Also takes an `OperationOptions` object.

File: types/google-cloud__storage/index.d.ts [hunk 7]
```diff
@@ -5,7 +5,340 @@
...
+        destination?: string;
+    }
+
+    /**
+     * Options when reading a file stream.
+     */
+    interface ReadStreamOptions {
+        end?: number;
+        start?: number;
+        validation?: string | boolean;
```
Comment (ento) [inline types/google-cloud__storage/index.d.ts]: Also takes a `userProject?: boolean` option.

File: types/google-cloud__storage/index.d.ts [hunk 8]
```diff
@@ -5,7 +5,340 @@
...
+    }
+
+    /**
+     * A file in the cloud.
+     */
+    class File {
+        constructor(bucket: Bucket, name: string, options?: BucketFileOptions);
+        acl: Acl;
+        copy(destination: string | Bucket | File): Promise<[File, ApiResponse]>;
+        createReadStream(options?: ReadStreamOptions): ReadStream;
```
Comment (ento) [inline types/google-cloud__storage/index.d.ts]: [Documentation](https://googlecloudplatform.github.io/google-cloud-node/#/docs/storage/1.2.0/storage/file?method=createReadStream) says the return value is a `ReadableStream`, not `fs.ReadStream`. May not make a difference in practice?

Same for [`createWriteStream`](https://googlecloudplatform.github.io/google-cloud-node/#/docs/storage/1.2.0/storage/file?method=createWriteStream) <-> `WritableStream`.

File: types/google-cloud__storage/index.d.ts [hunk 9]
```diff
@@ -5,7 +5,340 @@
...
+        createWriteStream(options?: WriteStreamOptions): WriteStream;
```
Comment (ento) [inline types/google-cloud__storage/index.d.ts]: [`createResumableUpload`](https://googlecloudplatform.github.io/google-cloud-node/#/docs/storage/1.2.0/storage/file?method=createResumableUpload) seems to be missing.

File: types/google-cloud__storage/index.d.ts [hunk 10]
```diff
@@ -5,7 +5,340 @@
...
+    interface WriteStreamOptions {
+        gzip?: boolean;
+        metadata?: FileMetadata;
+        offset?: number;
+        predefinedAcl?: string;
+        private?: boolean;
+        public?: boolean;
+        resumable?: boolean;
+        uri?: string;
+        validation?: string | boolean;
```
Comment (ento) [inline types/google-cloud__storage/index.d.ts]: Also takes a `userProject?: boolean` option.

File: types/google-cloud__storage/index.d.ts [hunk 11]
```diff
@@ -5,7 +5,340 @@
...
+    /**
+     * A file in the cloud.
+     */
+    class File {
+        constructor(bucket: Bucket, name: string, options?: BucketFileOptions);
+        acl: Acl;
+        copy(destination: string | Bucket | File): Promise<[File, ApiResponse]>;
+        createReadStream(options?: ReadStreamOptions): ReadStream;
+        createWriteStream(options?: WriteStreamOptions): WriteStream;
+        delete(): Promise<[ApiResponse]>;
```
Comment (ento) [inline types/google-cloud__storage/index.d.ts]: Also takes an `OperationOptions` object.

File: types/google-cloud__storage/index.d.ts [hunk 12]
```diff
@@ -5,7 +5,340 @@
...
+        download(options?: DownloadOptions): Promise<[Buffer]>;
+        exists(): Promise<[boolean]>;
```
Comment (ento) [inline types/google-cloud__storage/index.d.ts]: Also takes an `OperationOptions` object.

File: types/google-cloud__storage/index.d.ts [hunk 13]
```diff
@@ -5,7 +5,340 @@
...
+        get(): Promise<[File, ApiResponse]>;
```
Comment (ento) [inline types/google-cloud__storage/index.d.ts]: Also takes an `OperationOptions` object.

File: types/google-cloud__storage/index.d.ts [hunk 14]
```diff
@@ -5,7 +5,340 @@
...
+        getMetadata(): Promise<[FileMetadata, ApiResponse]>;
```
Comment (ento) [inline types/google-cloud__storage/index.d.ts]: Also takes an `OperationOptions` object.

File: types/google-cloud__storage/index.d.ts [hunk 15]
```diff
@@ -5,7 +5,340 @@
...
+
+    /**
+     * Options when obtaining the signed policy.
+     */
+    interface SignedPolicyOptions {
+        acl?: string;
+        contentLengthRange?: ContentLengthRange;
+        equals?: string[] | [string[]];
+        expires?: number | string;
+        max?: number;
```
Comment (ento) [inline types/google-cloud__storage/index.d.ts]: `max` and `min` are part of [contentLengthRange](https://googlecloudplatform.github.io/google-cloud-node/#/docs/storage/1.2.0/storage/file?method=getSignedPolicy): I think they can be removed from `SignedPolicyOptions`

File: types/google-cloud__storage/index.d.ts [hunk 16]
```diff
@@ -5,7 +5,340 @@
...
+     */
+    interface FileMetadata {
+        contentType?: string;
+    }
+
+    /**
+     * Options when setting the file as private.
+     */
+    interface FilePrivateOptions {
+        strict?: boolean;
```
Comment (ento) [inline types/google-cloud__storage/index.d.ts]: Also takes a `userProject?: boolean` option.

File: types/google-cloud__storage/index.d.ts [hunk 17]
```diff
@@ -5,7 +5,340 @@
...
+        delete(): Promise<[ApiResponse]>;
+        download(options?: DownloadOptions): Promise<[Buffer]>;
+        exists(): Promise<[boolean]>;
+        get(): Promise<[File, ApiResponse]>;
+        getMetadata(): Promise<[FileMetadata, ApiResponse]>;
+        getSignedPolicy(options?: SignedPolicyOptions): Promise<[SignedPolicy]>;
+        getSignedUrl(config?: SignedUrlConfig): Promise<[string]>;
+        makePrivate(options?: FilePrivateOptions): Promise<[ApiResponse]>;
+        makePublic(): Promise<[ApiResponse]>;
+        move(destination: string | Bucket | File): Promise<[File, ApiResponse]>;
```
Comment (ento) [inline types/google-cloud__storage/index.d.ts]: Also takes an `OperationOptions` object.

File: types/google-cloud__storage/index.d.ts [hunk 18]
```diff
@@ -5,7 +5,340 @@
...
+        name: string;
+        save(data: string, options?: WriteStreamOptions): Promise<void>;
+        setEncryptionKey(encryptionKey: string | Buffer): File;
+        setMetadata(metadata: FileMetadata): Promise<[ApiResponse]>;
```
Comment (ento) [inline types/google-cloud__storage/index.d.ts]: Also takes an `OperationOptions` object.

File: types/google-cloud__storage/index.d.ts [hunk 19]
```diff
@@ -5,7 +5,340 @@
...
+        metadata?: FileMetadata;
```
Comment (ento) [inline types/google-cloud__storage/index.d.ts]: [`setStorageClass`](https://googlecloudplatform.github.io/google-cloud-node/#/docs/storage/1.2.0/storage/file?method=setStorageClass) seems to be missing.

**Review by ento — APPROVED** [review]
Went through the official doc and noted any differences I saw: majority of them are from a previous lack of `requesterPays` support.

As an incremental improvement, I'm happy for this to get merged as-is, and treat my notes as to-dos to come back later.

**Comment by baronfel** [comment]
I agree with all of your notes, @ento, and will address them this morning. I will note, however, that my PR initially was a big copy/paste job with the main change being the fixing of the export functions.  All of your notes already existed as problems before my PR.

**Comment by sheetalkamat** [comment]
Merging this change as per reviewer's comment. @baronfel feel free to raise new PR for the other changes.

**Comment by ento** [comment]
@baronfel Yup, I saw the big copy/paste but also, as a reviewer, it felt not right to ignore discrepancies once I spotted one :) Thanks for fixing the import!

**Pull request merged**